### ISSUE TITLE
AOT scanner does absolutely nothing

### ISSUE THREAD

**Issue by Hertzole** [TeamSirenix/odin-serializer#38]
Hello

I've been fighting with AOT generation for several hours now and now it just feels like the AOT scanner is just broken.
I have a simple struct that I want to serialize. The struct itself also contains a few other structs. Nothing super advanced. It's all just ints, strings, etc. 
I have to add every single struct I want to serialize manually when generating an AOT dll. Is this really how it's supposed to be? I even had to add UnityEngine.Vector3Int!

This is my code for generating the dll file.
```cs
public void OnPreprocessBuild(BuildReport report)
{
    if (AOTSupportUtilities.ScanProjectForSerializedTypes(out List<Type> supportedTypes))
    {
        supportedTypes.Add(typeof(SaveData));
        supportedTypes.Add(typeof(ObjectData));
        supportedTypes.Add(typeof(ObjectData[]));
        supportedTypes.Add(typeof(ComponentData));
        supportedTypes.Add(typeof(ComponentData[]));
        supportedTypes.Add(typeof(PropertyData));
        supportedTypes.Add(typeof(PropertyData[]));
        supportedTypes.Add(typeof(Vector2Int));
        supportedTypes.Add(typeof(Vector3Int));

        AOTSupportUtilities.GenerateDLL(Application.dataPath + "/Plugins/", "GeneratedAOT", supportedTypes);
    }
}
```
Am I missing something here or is this by design?

I'm also sorry if this comes up as aggressive, I'm just really tired after working with this for several hours. I'm loving the rest though when it actually works in the editor and I don't want to give up on it just yet!

EDIT: I forgot to mention, checking the generated dll in an IL inspector shows that it's pretty much empty unless I manually add all this stuff.

**Comment by TorVestergaard** [comment]
What the scanner can catch very much depends on how it is saved in the project. The scanner does not scan assemblies or serialization code for usage, it does no code analysis (this would be immensely complicated to implement), it can *only* scan ScriptableObject assets, and Components on Prefab assets and in scenes - furthermore, it only scans things that it can prove are being included in the build, and which implement the ISerializationCallbackReceiver interface (IE, have custom serialization). It gets the scenes that are included in the build, scans those, and then uses the AssetDatabase to builds a dependency graph. It also scans the Resources and Asset Bundles, and various other things that can be proved to be included in the build.

As you might surmise, this means that the scan comes with a lot of asterisks and caveats attached - it will work fine for you in common usage where you merely derive from SerializedScriptableObject, SerializedMonoBehaviour, and so on. Anything where you invoke the serialization APIs yourself or do anything custom will not (*cannot*) be scanned by default, and you will have to instruct the scan to catch all those types by yourself, either by being clever and automating the location of your own required types (somehow), or by manually adding them all as you're currently doing.

**Comment by pshtif** [comment]
Tor you mention in your reply that it scans "Components on Prefab assets and in scenes" however it doesn't seem to be scanning components on Prefabs that are not already in the scene and are added later using Instatiate. Or am I missing something? It just goes over assets and checks if they directly implement ISerializationCallbackReceiver which they don't but one of their children has components that do. Which gets ignored by the scan.

So basically a prefab asset is scanned using just this:

`public void ScanObject(UnityEngine.Object obj)
        {
            if (obj is ISerializationCallbackReceiver)
            {
...
            }
        }`

Which immediately discards it as the prefab itself doesn't implement the ISerializationCallbackReceiver only his children do.

**Comment by TorVestergaard** [comment]
Prefabs are scanned via ScanAsset, which calls AssetDatabase.LoadAllAssetsAtPath(assetPath) and then scans all of the instances which are returned by Unity as being contained in that asset - which should include all components.
